File: spinoza/examples/expectation.rs
use clap::Parser;
use spinoza::core::qubit_expectation_value;
use spinoza::{
    config::{Config, QSArgs},
    core::{xyz_expectation_value, State, CONFIG},
    gates::{apply, Gate},
};

fn main() {
    let args = QSArgs::parse();
    let config = Config::from_cli(args);
    CONFIG.set(config).unwrap();

    let n = config.qubits.into();
    let mut state = State::new(n);
    let target = 0;

    apply(Gate::RX(0.54), &mut state, target);
    apply(Gate::RY(0.12), &mut state, target);

    let targets = (0..n).collect::<Vec<usize>>();
    let exp_vals = xyz_expectation_value('z', &state, &targets);
    println!("expectation values: {:?}", exp_vals);

    let exp_vals: Vec<_> = (0..n).map(|t| qubit_expectation_value(&state, t)).collect();
    println!(
        "expectation values using `qubit_expectation_value`: {:?}",
        exp_vals
    );
}

File: spynoza/examples/expectation.py
from spynoza import QuantumCircuit, QuantumRegister, qubit_expectation_value, xyz_expectation_value


def main():
    n = 3
    qr = QuantumRegister(n)
    qc = QuantumCircuit(qr)
    target = 0

    qc.rx(0.54, target)
    qc.ry(0.12, target)
    qc.execute()

    targets = list((range(n)))
    exp_vals = xyz_expectation_value('z', qc.get_statevector(), targets)
    print(f"expectation values: {exp_vals}")

    exp_vals = [None] * 3
    for i, target in enumerate(targets):
        exp_vals[i] = qubit_expectation_value(qc.get_statevector(), target)

    print(f"expectation values using `qubit_expectation_value`: {exp_vals}")


if __name__ == "__main__":
    main()